gateway-discovery-cm/src/main/java/org/apache/knox/gateway/topology/discovery/cm/ClouderaManagerClusterConfigurationMonitor.java

```diff
@@ -0,0 +1,940 @@
+                    // the "new" service, but discovery had previously not succeeded because the service had not been
+                    // configured (appropriately) at that time.
+                    log.serviceEnabled(re.getService());
+                    configHasChanged = true;
+                  }
+
+                  handledServiceTypes.add(serviceType);
+                }
+
+                if (configHasChanged) {
+                  break; // No need to continue checking once we've identified one reason to perform discovery again
+                }
+              }
+
+              // If a change has occurred, notify the listeners
+              if (configHasChanged) {
+                monitor.notifyChangeListeners(address, clusterName);
+              }
+            }
+          }
+        }
+
+        waitFor(interval);
+      }
+    }
+  }
+
+
+  /**
+   * Internal representation of a ClouderaManager service restart event
+   */
+  private static final class RestartEvent {
+
+    private static final String ATTR_CLUSTER      = "CLUSTER";
+    private static final String ATTR_SERVICE_TYPE = "SERVICE_TYPE";
+    private static final String ATTR_SERVICE      = "SERVICE";
+
+    private static List<String> attrsOfInterest = new ArrayList<>();
+    static {
+      attrsOfInterest.add(ATTR_CLUSTER);
+      attrsOfInterest.add(ATTR_SERVICE_TYPE);
+      attrsOfInterest.add(ATTR_SERVICE);
+    }
+
+    private ApiEvent auditEvent;
+    private String   clusterName;
+    private String   serviceType;
+    private String   service;
+
+    RestartEvent(final ApiEvent auditEvent) {
+      if (ApiEventCategory.AUDIT_EVENT != auditEvent.getCategory()) {
+        throw new IllegalArgumentException("Invalid event category " + auditEvent.getCategory().getValue());
+      }
+      this.auditEvent = auditEvent;
+      for (ApiEventAttribute attribute : auditEvent.getAttributes()) {
+        if (attrsOfInterest.contains(attribute.getName())) {
+          setPropertyFromAttribute(attribute);
+        }
+      }
+    }
+
+    String getTimestamp() {
+      return auditEvent.getTimeOccurred();
+    }
+
+    String getClusterName() {
+      return clusterName;
+    }
+
+    String getServiceType() {
+      return serviceType;
+    }
+
+    String getService() {
+      return service;
+    }
+
+    private void setPropertyFromAttribute(ApiEventAttribute attribute) {
+      switch (attribute.getName()) {
+        case ATTR_CLUSTER:
+          clusterName = attribute.getValues().get(0);
+          break;
+        case ATTR_SERVICE_TYPE:
+          serviceType = attribute.getValues().get(0);
+          break;
+        case ATTR_SERVICE:
+          service = attribute.getValues().get(0);
+          break;
+        default:
+      }
+    }
+  }
+
+
+  /**
+   * Service configuration properties container
+   */
+  private static final class ServiceConfigurationProperties implements Serializable {
```

Comment:
I don't think we should have a serializable class. Maybe a jackson pojo for json?